compiler/rustc_parse/src/lexer/mod.rs

```diff
@@ -203,6 +208,58 @@ impl<'a> StringReader<'a> {
                         .push(span);
                     token::Ident(sym, false)
                 }
+                // split up (raw) c string literals to an ident and a string literal when edition < 2021.
+                rustc_lexer::TokenKind::Literal {
+                    kind: kind @ (LiteralKind::CStr { .. } | LiteralKind::RawCStr { .. }),
+                    suffix_start,
+                } if !self.mk_sp(start, self.pos).edition().rust_2021() => {
+                    let lit_start = start + BytePos(match kind {
+                        LiteralKind::CStr { .. } => 1,
+                        LiteralKind::RawCStr { .. } => 2,
+                        _ => unreachable!(),
+                    });
+
+                    let suffix_start = start + BytePos(suffix_start);
+
+                    let (kind, symbol) = self.cook_lexer_literal(
+                        lit_start,
+                        suffix_start,
+                        match kind {
+                            rustc_lexer::LiteralKind::CStr { terminated } => rustc_lexer::LiteralKind::Str { terminated },
+                            rustc_lexer::LiteralKind::RawCStr { n_hashes } => rustc_lexer::LiteralKind::RawStr { n_hashes },
+
+                            #[cfg(debug_assertions)]
+                            _ => unreachable!(),
+
+                            #[cfg(not(debug_assertions))]
+                            x => x,
+                        }
+                    );
+
+                    let suffix = if suffix_start < self.pos {
+                        let string = self.str_from(suffix_start);
+                        if string == "_" {
+                            self.sess
+                                .span_diagnostic
+                                .emit_err(errors::UnderscoreLiteralSuffix { span: self.mk_sp(suffix_start, self.pos) });
+                            None
+                        } else {
+                            Some(Symbol::intern(string))
+                        }
+                    } else {
+                        None
+                    };
+                    self.override_next = Some(Token::new(token::Literal(token::Lit { kind, symbol, suffix }), self.mk_sp(lit_start, self.pos)));
```

Comment:
Can't we set `cursor`, `pos` and whatever else is necessary to get this literal naturally on the next call to `next_token`?
(Instead of constructing it manually from `LiteralKind::(Raw)CStr`.)

Reply:
done